### PR TITLE
Added implementation for Bezier Curve, under a new graphics directory. 

### PR DESCRIPTION
**Pull request by fk03983**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.

### CODE REVIEW

File: graphics/bezier_curve.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# https://en.wikipedia.org/wiki/B%C3%A9zier_curve
+# https://www.tutorialspoint.com/computer_graphics/computer_graphics_curves.htm
+
+from typing import List, Tuple
+import scipy.special as sp
```
Comment (cclauss) [inline graphics/bezier_curve.py]: from scipy.special import comb

File: graphics/bezier_curve.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class BezierCurve:
+    """
+    A class to generate bezier curves for a given set of points. The implementation works only for 2d points.
```
Comment (cclauss) [inline graphics/bezier_curve.py]:
```
Generate bezier curves for a given set of points.
This implementation works only for 2d points.
```
Lines should not be longer than 88 characters.

File: graphics/bezier_curve.py (hunk 1 above)
Comment (fk03983) [reply graphics/bezier_curve.py]: I have edited it.

File: graphics/bezier_curve.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    """
+
+    def __init__(self, list_of_points: List[Tuple]):
```
Comment (cclauss) [inline graphics/bezier_curve.py]: List[Tuple[float, float]]. # if that works

File: graphics/bezier_curve.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        """
+        The constructor of the Bezier Curve class.
+            list_of_points: List of tuples on which to interpolate.
```
Comment (cclauss) [inline graphics/bezier_curve.py]: list_of_points: List of tuples __of two floats__ on which to interpolate.

In general, the type annotations describe the data types so we do not need to repeat that in our comments.  The comments should focus on describing what the data represents or why it is important.

File: graphics/bezier_curve.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            returns a returns list of floating point numbers, a number at index i represents the value of
+            basis function i at t.
+
+        >>> curve1 = BezierCurve([(1,1), (1,2)])
+        >>> curve1.basis_function(0)
+        [1.0, 0.0]
+        >>> curve1.basis_function(1)
+        [0.0, 1.0]
+        """
+        output_values: List[float] = list()
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Should we `assert 0 <= t <= 1` ?

We do not need type annotations outside of the function signatures.  Let's do this a list comprehension...
```
x_y = tuple([comb(self.degree, i) * ((1 - t) ** (self.degree - i)) * (t ** i)
             for i in range(len(self.list_of_points))])
```

File: graphics/bezier_curve.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        for i in range(len(self.list_of_points)):
+            output_values.append(
+                sp.comb(self.degree, i) * ((1 - t) ** (self.degree - i)) * (t ** i)
+            )  # basis function for each i
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Put the comment on the line before or remove the comment.

File: graphics/bezier_curve.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        assert (
+            round(sum(output_values), 5) == 1
+        )  # the basis must sum up to 1 for it to produce a valid bezier curve.
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Put the comment on the line before or remove the comment.

File: graphics/bezier_curve.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        (1.0, 1.0)
+        >>> curve1.bezier_curve_function(1)
+        (1.0, 2.0)
+        """
+        basis_function = self.basis_function(t)
+        x = 0.0
+        y = 0.0
+        for i in range(
+            len(self.list_of_points)
+        ):  # summing up the product of i-th basis function and ith point, for all points
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Put the comment on the line before or remove the comment.

File: graphics/bezier_curve.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    def __init__(self, list_of_points: List[Tuple]):
+        """
+        The constructor of the Bezier Curve class.
+            list_of_points: List of tuples on which to interpolate.
+        """
+        self.list_of_points = list_of_points
+        self.degree = (
+            len(list_of_points) - 1
+        )  # degree of the curve produced will be n - 1
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Please move the comment to the line above or remove the comment.  For me (who knows nothing about Bezier curves) the comment would tell me _why_ the degree is important or _how_ we will use it.  The code already tells me `n - 1`.

File: graphics/bezier_curve.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    def basis_function(self, t: float) -> List[float]:
+        """
+        The function to cater the need for the basis function.
+            t: the 1-d co-ordinate at which to evaluate the basis function(s).
+            returns a returns list of floating point numbers, a number at index i represents the value of
+            basis function i at t.
```
Comment (cclauss) [inline graphics/bezier_curve.py]: This comment is really unclear.  Please clearly define what basis is.
t: time value between 0 and 1 inclusive at which to evaluate the basis of the curve.
returns the x, y values of basis function at time t

Wrap lines at 88 characters.

File: graphics/bezier_curve.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        it must return the first point, and at t = 1, it must return the last point.
+            t: the value of t at which to evaluate the bezier function
+        returns a 2-d tuple of floats, representing the value of the bezier curve at the given t.
+
+        >>> curve1 = BezierCurve([(1,1), (1,2)])
+        >>> curve1.bezier_curve_function(0)
+        (1.0, 1.0)
+        >>> curve1.bezier_curve_function(1)
+        (1.0, 2.0)
+        """
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Should we `assert 0 <= t <= 1` ?

File: graphics/bezier_curve.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            round(sum(output_values), 5) == 1
+        )  # the basis must sum up to 1 for it to produce a valid bezier curve.
+        return output_values
+
+    def bezier_curve_function(self, t: float) -> Tuple[float, float]:
+        """
+        The function to produce the values of the bezier curve at time t. At t = 0,
+        it must return the first point, and at t = 1, it must return the last point.
+            t: the value of t at which to evaluate the bezier function
+        returns a 2-d tuple of floats, representing the value of the bezier curve at the given t.
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Returns the x, y coordinates of the Bezier curve at time t.
The first point in the curve is when t = 0.
The last point in the curve is when t = 1.

File: graphics/bezier_curve.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        ):  # summing up the product of i-th basis function and ith point, for all points
+            x += basis_function[i] * self.list_of_points[i][0]
+            y += basis_function[i] * self.list_of_points[i][1]
+
+        return (x, y)
+
+    def plot_curve(self, step_size: float = 0.01):
+        """
+        plots the bezier curve using matplotlib plotting function.
+            step_size: defines the step(s) at which to evaluate the bezier curve and plot. The smaller the step size,
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Wrap lines at 88 characters.

File: graphics/bezier_curve.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        plt.show()
+
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    curve1 = BezierCurve([(1, 2), (3, 5)])
+    curve1.plot_curve()
```
Comment (cclauss) [inline graphics/bezier_curve.py]: No need to create a variable here...
`BezierCurve([(1, 2), (3, 5)]).plot_curve()`

File: graphics/bezier_curve.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+        (1.0, 1.0)
+        >>> curve1.bezier_curve_function(1)
+        (1.0, 2.0)
+        """
+        basis_function = self.basis_function(t)
+        x = 0.0
+        y = 0.0
+        for i in range(
+            len(self.list_of_points)
+        ):  # summing up the product of i-th basis function and ith point, for all points
```
Comment (cclauss) [inline graphics/bezier_curve.py]: Change the second ìth` to ì-th`to placate the `codespell` test.

File: graphics/bezier_curve.py (hunk 5 above)
Comment (fk03983) [reply graphics/bezier_curve.py]: Agreed with the assert part. Don't you think doing this as a list comprehension reduces the readability as there is a lot going on inside?

**Comment by fk03983** [comment]
Thanks! Look forward to making more of these contributions!

**Pull request merged**